pkg/controller/controller.go

```diff
@@ -17,196 +17,315 @@ limitations under the License.
 
-	return eStore
+	wait.Until(c.runWorker, time.Second, stopCh)
 }
 
-func watchJobs(client *client.ExtensionsClient, eventHandler handlers.Handler) cache.Store {
-	//Define what we want to look for (Jobs)
-	watchlist := cache.NewListWatchFromClient(client, "jobs", api.NamespaceAll, fields.Everything())
+// HasSynced is required for the cache.Controller interface.
+func (c *Controller) HasSynced() bool {
+	return c.informer.HasSynced()
+}
 
-	resyncPeriod := 30 * time.Minute
+// LastSyncResourceVersion is required for the cache.Controller interface.
+func (c *Controller) LastSyncResourceVersion() string {
+	return c.informer.LastSyncResourceVersion()
+}
 
-	//Setup an informer to call functions when the watchlist changes
-	eStore, eController := framework.NewInformer(
-		watchlist,
-		&v1beta1.Job{},
-		resyncPeriod,
-		framework.ResourceEventHandlerFuncs{
-			AddFunc:    eventHandler.ObjectCreated,
-			DeleteFunc: eventHandler.ObjectDeleted,
-		},
-	)
+func (c *Controller) runWorker() {
+	for c.processNextItem() {
+		// continue looping
+	}
+}
 
-	//Run the controller as a goroutine
-	go eController.Run(wait.NeverStop)
+func (c *Controller) processNextItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.processItem(key.(string))
+	if err == nil {
+		// No error, reset the ratelimit counters
+		c.queue.Forget(key)
+	} else if c.queue.NumRequeues(key) < maxRetries {
+		c.logger.Errorf("Error processing %s (will retry): %v", key, err)
+		c.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		c.logger.Errorf("Error processing %s (giving up): %v", key, err)
+		c.queue.Forget(key)
+		utilruntime.HandleError(err)
+	}
 
-	return eStore
+	return true
 }
 
-func watchPersistenVolumes(client *client.Client, store cache.Store, eventHandler handlers.Handler) cache.Store {
-	//Define what we want to look for (PersistenVolumes)
-	watchlist := cache.NewListWatchFromClient(client, "persistentvolumes", api.NamespaceAll, fields.Everything())
-
-	resyncPeriod := 30 * time.Minute
+func (c *Controller) processItem(key string) error {
+	c.logger.Infof("Processing change to Pod %s", key)
 
-	//Setup an informer to call functions when the watchlist changes
-	eStore, eController := framework.NewInformer(
-		watchlist,
-		&api.PersistentVolume{},
-		resyncPeriod,
-		framework.ResourceEventHandlerFuncs{
-			AddFunc:    eventHandler.ObjectCreated,
-			DeleteFunc: eventHandler.ObjectDeleted,
-		},
-	)
+	obj, exists, err := c.informer.GetIndexer().GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
+	}
 
-	//Run the controller as a goroutine
-	go eController.Run(wait.NeverStop)
+	if !exists {
+		c.eventHandler.ObjectDeleted(obj)
+		return nil
+	}
 
-	return eStore
+	c.eventHandler.ObjectCreated(obj)
+	return nil
 }
+
+//
+//func watchServices(client *client.Client, eventHandler handlers.Handler) cache.Store {
```

Comment:
There will be separate controller for each resource. I only use one for Pod in this PR mostly for showing use-case for the 2nd part of the blogpost. I will cover the rest in another PR